### ISSUE TITLE
Add option to restrict front-/back-channel logout to RPs associated with the logged out session

### ISSUE THREAD

**Issue by tyaps** [ory/hydra#1660]
**Describe the bug**

If user was previously logged in multiple RPs during some OIDC sessions, when user does logout from current OIDC session, it gets frontChannel logout script with all of RP's URLs, even it was logged to only one RP during current OIDC session.

While doing session logout, when preparing frontChannelUris, sql query uses only subject, but doesn't use sid. 

func (s *DefaultStrategy) generateFrontChannelLogoutURLs(ctx context.Context, subject, sid string) ([]string, error) {
	clients, err := s.r.ConsentManager().ListUserAuthenticatedClientsWithFrontChannelLogout(ctx, subject)

**Reproducing the bug**

Login to several RP's (RP1, RP2) during one OIDC session (remember=true).
Logout this OIDC session.
Login to one RP1 during new OIDC session (remember=true).
Logout this OIDC session.
frontChannel logout script will contain frontChannelLogout URLs list for RP1 and RP2

**Expected behavior**

 frontChannel logout script should contain frontChannelLogout URLs only for RP1

**Comment by aeneasr** [comment]
I believe this is intentional, although I'm not sure. Do you have a link/resource to a spec or implementation guideline where this behavior is defined?

**Comment by tyaps** [comment]
I'll try to find it.
But from common sense I think, that if you were logged in during one OIDC session to RP1, then if you do logout, only that RP1 should be notified. Why some other RPs should know that you logged out...
Ok, I'll search some evidences ))

**Comment by aeneasr** [comment]
I believe that we send the `sid` with the request, that way, the RP can decide if they wish to log the user out based on that sid or not.

**Comment by tyaps** [comment]
Hmm... by the way about sid. I thought that RP doesn't know anything about sid. Sid (OIDC sessionID)  never arrives on RP: when you change code on token there is no any Sid parameters in queryString. RP can't get it somehow.

Sid (that is represented with oauth2_authentication_session cookie) is only between User and OIDC Provider to remember successful login occurrence. Or do I miss something.

(haven't got time to search guidelines, but I'll find some time to investigate soon)

**Comment by aeneasr** [comment]
https://stackoverflow.com/questions/45370324/what-is-sid-claim-in-logout-token-in-openid-connect-back-channel-logout

**Comment by tyaps** [comment]
Ok, thank you. I can get sid from id_token.

But here is still my considerations.
I examined https://openid.net/specs/openid-connect-frontchannel-1_0.html more thoroughly and have  the following:

1. about using sid.

> frontchannel_logout_session_required
OPTIONAL. Boolean value specifying whether the RP requires that iss (issuer) and sid (session ID) query parameters be included to identify the RP session with the OP when the frontchannel_logout_uri is used. If omitted, the default value is false.

For me it means that sid in frontChannelLogout request is optional  and in theory frontChannelLogout can handle without it. I mean, in general way we mustn't rely on this parameter as an obligatory. 

2. 

> OpenID Provider Logout Functionality
OPs supporting HTTP-based logout need to keep track of the set of logged-in RPs so that they know what RPs to contact at their logout URIs to cause them to log out. 

I think it is a clear and direct sentence, that means that OP must send frontChannelLogouts only to RPs that were involved in OIDC session.

3.

Again, imagine the scenario, how it happens with current approach:

Suppose we have new user, that was newer logged in before. It does the following:

```
Stage 1.
Login to RP1.
Logout.
result: frontChannelLogout request is sent to RP1.

Stage 2.
Login to RP1.
Login to RP2.
Logout.
result: frontChannelLogout request is sent to RP1 and RP2.

Stage 3.
Login to RP1.
Logout.
result: frontChannelLogout request is sent to RP1 and RP2.

The end.
```

For me Stage 3 should behave like Stage 1 because (first of all) it is the same story and it shouldn't depend on previous login events.

Or Stage 1 should behave like Stage 3 (but it is completely impossible, because we can't send frontChannelLogout to all OIDC-known RPs)

So in summary I still consider this behaviour as a bug.

**Comment by aeneasr** [comment]
Sorry for the late reply, this slipped my radar.

I checked some other servers, https://connect2id.com/blog/connect2id-server-7-3 seems to be implementing it as you suggest:

> Implements OpenID Connect front-channel logout 1.0 (draft 02) for notifying subscribed OpenID relying parties when a **subject session into which the OpenID relying party was logged in gets closed**. The notifications are delivered via the browser, by calling a registered OpenID relying party URL (typically by the OpenID provider creating a hidden iframe for the URL). Front-channel logout is implemented by extending the Connect2id server logout session web API.

as well as:

> Extends the subject (end-user) session store API to track OpenID relying parties **logged in during a session**.

However, I'm still trying to find official guidance. I might just reach out to the OIDC foundation to figure this out. Definitely appears to be more on the bug side now, though!

**Comment by aeneasr** [comment]
So I've sent an email to the mailing list. Here's the thread:

> Hi Aeneas,
>
> The specifications say the OP should keep track of the “visited sites” / RPs so that, when logout notifications go out it knows which ones to contact. 
>
> > OPs supporting HTTP-based logout need to keep track of the set of logged-in RPs so that they know what RPs to contact at their logout URIs to cause them to log out. Some OPs track this state using a "visited sites" cookie.
>
> But I don’t believe it also forbids contacting all of them, though i believe that doesn’t scale well. 


I replied:

> thank you for the fast response and guidance! I saw that section but I understood it as: We need to keep track of the active RPs session (regardless of the user agent/specific session) to contact them all. Maybe this could be refined with an addition such as:
>
> OPs SHOULD (or MAY) only contact the RPs associated with the user session that is being logged out.
>
> Thank you for clarifying that both options are viable (although one may not scale very well). Are there any other implications this could have?

He replied:

> Personally I don’t believe removing all end user RP sessions is the expected behaviour. For one, front channel logout notification is not likely to do much. 
>
> I can only see a handful of scenarios where me logging out of a website at work should log me out of all RPs on my phone and home office. 
>
> Maybe allow your library developers to choose the behaviour they want but the default expected one from my POV is only clearing RP sessions i visited in the session i’m at. 

TLDR;

While it may be difficult to scale and may not be the expected behavior by everyone, both options are valid. We can however either define this to be configurable, or enforce one or the other mechanism.

As a workaround I would expect the RP (relying party/client) to check the SID before logging out!

**Comment by panva** [comment]
> As a workaround I would expect the RP (relying party/client) to check the SID before logging out!

That's an effort by each individual RP, an unexpected one.

There's a difference between the specification explicitly not forbidding something and making it a valid option. One can simply not forbid everything readers may interpret in a specification. **The intent of the specification is clear, only notify RPs from the current session that logout is occuring in.**

**Comment by aeneasr** [comment]
Thank you for the clarification! Set to 1.1.1 milestone :)
